content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md

```diff
@@ -0,0 +1,127 @@
+
+```yaml
+capabilities:
+  drop:
+  - ALL
+  add:
+  - NET_BIND_SERVICE
+  - SYS_CHROOT
+```
+
+If you use the Helm Chart, in your `values.yaml`, change the following:
+
+```yaml
+controller:
+  image:
+    chroot: true
+```
+
+** Check with your cluster admin if you can use "SYS_CHROOT" capability before enabling it in your deployment! **
+
+## OK, but how does this increase the security of my Ingress Controller?
+
+Take the following configuration snippet and imagine, for some reason it was added to your nginx.conf:
+```
+location /randomthing/ {
+      alias /;
+      autoindex on;
+}
+```
+
+With this configuration, one can call "http://website.com/randomthing" and get some listing (and access) to the whole filesystem.
+
+Now, can you spot the difference between chrooted and non chrooted Nginx on the listings below?
+
+| not chrooted             | chrooted |
+|--------------------------|------|
+| bin                      | bin  |
+| dev                      | dev  |
+| etc                      | etc  |
+| home                     |      |
+| lib                      | lib  |
+| media                    |      |
+| mnt                      |      |
+| opt                      | opt  |
+| proc                     | proc |
+| root                     |      |
+| run                      | run  |
+| sbin                     |      |
+| srv                      |      |
+| sys                      |      |
+| tmp                      | tmp  |
+| usr                      | usr  |
+| var                      | var  |
+| dbg                      |      |
+| nginx-ingress-controller |      |
+| wait-shutdown            |      |
+
+The one in left side is not chrooted. So NGINX has full access to the filesystem. The one in right side is chrooted, so a new filesystem with only the required files to make NGINX work is created.
+
+## What about other security improvements in this release?
+
+We know that the chroot improvement solves some portion of the risk, but still, someone can try to inject commands to read, for example, nginx.conf file and extract sensitive information.
+
+So, another change in this release (this is opt-out!) is the deep inspector. We know some directives and regexes may be dangerous to NGINX, so the deep inspector will check all fields from an Ingress object (during its reconciliation, and also with admission webhook!) to verify if some field contains these dangerous directives.
+
+We already do this for annotations, and the plan/goal is to move this validation to inside deep inspection in future releases.
+
+You can take a look into the existing rules in <https://github.com/kubernetes/ingress-nginx/blob/main/internal/ingress/inspector/rules.go>
+
+Due to the nature of inspecting and matching all strings in all objects, this new feature may consume a bit more CPU, and can be disabled with the flag --deep-inspect=false
```

Comment:
Is this right:
```suggestion
Due to the nature of inspecting and matching all strings within relevant Ingress objects, this new feature may consume a bit more CPU. You can disable it by running the ingress controller with the command line argument `--deep-inspect=false`.
```

Reply:
so, today the deep inspector does only Ingress objects but it's prepared for service objects (as an example, one may want to block localhost services for CVE reasons).

We may improve this tho, with your review but not discarding that other objects may be inspected

Reply:
actually right now it does make sense. Let me commit here